packages/integration-tests/src/retry.ts

```diff
@@ -0,0 +1,44 @@
+/**
+ * Executes an asynchronous operation with retry logic.
+ * This function attempts to execute the provided callable function multiple times
+ * based on the specified retry conditions. It's useful for handling transient
+ * errors or temporary service unavailability.
+ */
+export const runWithRetry = async <T>(
+  callable: () => Promise<T>,
+  retryPredicate: (error: Error) => boolean,
+  maxAttempts = 3,
+  timeout?: number
+): Promise<T> => {
+  const startTime = Date.now();
+  const collectedErrors: Error[] = [];
+
+  for (let attempt = 1; attempt <= maxAttempts; attempt++) {
+    try {
+      if (timeout && Date.now() - startTime >= timeout) {
+        throw new Error(`Operation timed out after ${timeout}ms`);
+      }
+
+      const result = await callable();
+      return result;
+    } catch (error) {
+      if (error instanceof Error) {
+        collectedErrors.push(error);
+
+        if (!retryPredicate(error) || attempt === maxAttempts) {
+          break;
+        }
+
+        console.log(`Attempt ${attempt} failed. Retrying...`);
```

Comment:
I'd remove this log.
It does not say which operation is being retried so logging that is going to be noise rather than useful information. We won't care much if operation succeeds within max attempts.

Throwing aggregate error at the end gives information about number of retires and how they failed.